src/Build/Graph/GraphBuilder.cs

```diff
@@ -161,28 +161,42 @@ private void AddEdgesFromProjectReferenceItems(Dictionary<ConfigurationMetadata,
             }
 
             HashSet<ProjectGraphNode> GetTransitiveProjectReferencesExcludingSelf(ParsedProject parsedProject)
+            {
+                HashSet<ProjectGraphNode> references = new();
+                GetTransitiveProjectReferencesExcludingSelfHelper(parsedProject, references, null);
+                return references;
+            }
+
+            void GetTransitiveProjectReferencesExcludingSelfHelper(ParsedProject parsedProject, HashSet<ProjectGraphNode> transitiveReferences, HashSet<ProjectGraphNode> referencesFromHere)
```

Comment:
I don't think I understand what `referencesFromHere` does. Is there another name? Below you call it with `toCache`.

Reply:
I'm open to better names.

transitiveReferences holds all references we have found so far.
toCache stores all references (including transitive references) from the project (parsedProject) being considered in one particular recursive call.
referencesFromHere is like toCache but from the parent project.

Let's say project A references project B and project C. Project B references project D, and D references E.
When B is "parsedProject," referencesFromHere contains all of A's references, so it either does (if C were first) or will (if B comes first) contain C, but C is not part of B's transitive reference tree, hence toCache. toCache, at line 183, is passed as referencesFromHere to the recursive call for project D. D will add both its references (through passing toCache in line 183 (transitive references) and adding direct references directly on line 189. At this point, toCache holds all of D's references exluding itself, so we add it to transitiveReferenceCache.

At that point, B's toCache (passed to D as referencesFromHere) has nothing. D's recursive call adds E to it on 195 by adding its own references. Then, in B's recursive call, it adds D on 189. Now it has all of its (direct and indirect) references.

Does that make sense? Recursion can be very confusing (and having a bug didn't help).

Reply:
```suggestion
            void GetTransitiveProjectReferencesExcludingSelfHelper(ParsedProject parsedProject, HashSet<ProjectGraphNode> traversedReferences, HashSet<ProjectGraphNode> referencesOfReferencingNode)
```

?

Reply:
referencesOfDirectlyReferencingNode?

I'm a little unsure about either name because it makes it sound like something the recursive calls could use rather than something incomplete to add to.

incompleteReferencesOfDirectlyReferencingNode?